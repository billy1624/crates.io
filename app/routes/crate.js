import { NotFoundError } from '@ember-data/adapter/error';
import Route from '@ember/routing/route';
import { inject as service } from '@ember/service';
import { ajax_fail, htmlDecode } from '../utils/ajax';

export default class CrateRoute extends Route {
  @service headData;
  @service router;
  @service store;

  async model(params, transition) {
    let crateName = params.crate_id;

    try {
      return (async () => {
        const data = await ajax_fail(
          `https://raw.githubusercontent.com/SeaQL/rustacean.info/main/related-articles/${crateName}.json`,
        );
        const res = await this.store.findRecord('crate', crateName);
        Object.assign(res, {
          articles: data.map(row => {
            row.title = htmlDecode(row.title);
            return row;
          }),
          link_contribute_articles: `https://github.com/SeaQL/rustacean.info/blob/main/CONTRIBUTING.md`,
          link_create_articles: `https://github.com/SeaQL/rustacean.info/new/main/related-articles`,
          link_edit_articles: `https://github.com/SeaQL/rustacean.info/edit/main/related-articles/${crateName}.json`,
        });
        return res;
      })();
    } catch (error) {
      if (error instanceof NotFoundError) {
        let title = `${crateName}: Crate not found`;
        this.router.replaceWith('catch-all', { transition, error, title });
      } else {
        let title = `${crateName}: Failed to load crate data`;
        this.router.replaceWith('catch-all', { transition, error, title, tryAgain: true });
      }
    }
  }

  setupController(controller, model) {
    super.setupController(...arguments);
    this.headData.crate = model;
  }

  resetController() {
    super.resetController(...arguments);
    this.headData.crate = null;
  }
}
