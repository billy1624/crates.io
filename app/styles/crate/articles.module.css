.results-meta {
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-bottom: var(--space-2xs);

    @media only screen and (max-width: 550px) {
        display: block;
    }
}

.page-description {
    composes: small from '../shared/typography.module.css';

    @media only screen and (max-width: 550px) {
        display: block;
        margin-bottom: var(--space-s);
    }
}
.list {
    list-style: none;
    margin: 0 0 var(--space-s);
    padding: 0;

    > * + * {
        margin-top: var(--space-2xs);
    }
}
