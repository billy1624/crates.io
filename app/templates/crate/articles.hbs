<CrateHeader @crate={{this.model}} />

<ul local-class="list">
  {{#if this.model.articles}}
    {{#each this.model.articles as |article|}}
      <li>
        <ArticleList::Row @link={{article.link}} @title={{article.title}} local-class="row" data-test-version={{article.title}} />
      </li>
    {{/each}}
    <br>
    Contribute and post more related articles on <a href={{this.model.link_edit_articles}} target="_blank">GitHub</a>!
  {{else}}
    Contribute and post the first related article on <a href={{this.model.link_create_articles}} target="_blank">GitHub</a>!
  {{/if}}
  Read contributing guide <a href={{this.model.link_contribute_articles}} target="_blank">here</a>.
</ul>
