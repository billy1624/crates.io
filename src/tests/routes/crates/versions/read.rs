use crate::builders::{CrateBuilder, VersionBuilder};
use crate::util::insta::{self, assert_json_snapshot};
use crate::util::{RequestHelper, TestApp};
use diesel::prelude::*;
use serde_json::Value;

#[tokio::test(flavor = "multi_thread")]
async fn show_by_crate_name_and_version() {
    let (app, anon, user) = TestApp::init().with_user();
    let user = user.as_model();

    let v = app.db(|conn| {
        let krate = CrateBuilder::new("foo_vers_show", user.id).expect_build(conn);
        VersionBuilder::new("2.0.0")
            .size(1234)
            .checksum("c241cd77c3723ccf1aa453f169ee60c0a888344da504bee0142adb859092acb4")
            .rust_version("1.64")
            .expect_build(krate.id, user.id, conn)
    });

    let url = "https://crates.io/api/v1/crates/foo_vers_show/2.0.0";
    let json: Value = anon.get(url).await.good();
    assert_json_snapshot!(json, {
        ".version.id" => insta::id_redaction(v.id),
        ".version.created_at" => "[datetime]",
        ".version.updated_at" => "[datetime]",
        ".version.published_by.id" => insta::id_redaction(user.id),
    });
}

#[tokio::test(flavor = "multi_thread")]
async fn show_by_crate_name_and_semver_no_published_by() {
    use crates_io::schema::versions;
    use diesel::{update, RunQueryDsl};

    let (app, anon, user) = TestApp::init().with_user();
    let user = user.as_model();

    let v = app.db(|conn| {
        let krate = CrateBuilder::new("foo_vers_show_no_pb", user.id).expect_build(conn);
        let version = VersionBuilder::new("1.0.0").expect_build(krate.id, user.id, conn);

        // Mimic a version published before we started recording who published versions
        let none: Option<i32> = None;
        update(versions::table)
            .set(versions::published_by.eq(none))
            .execute(conn)
            .unwrap();

        version
    });

    let url = "https://crates.io/api/v1/crates/foo_vers_show_no_pb/1.0.0";
    let json: Value = anon.get(url).await.good();
    assert_json_snapshot!(json, {
        ".version.id" => insta::id_redaction(v.id),
        ".version.created_at" => "[datetime]",
        ".version.updated_at" => "[datetime]",
    });
}
