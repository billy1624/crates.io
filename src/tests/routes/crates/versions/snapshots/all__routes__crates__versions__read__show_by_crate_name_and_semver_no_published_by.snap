---
source: src/tests/routes/crates/versions/read.rs
expression: json
---
{
  "version": {
    "audit_actions": [],
    "bin_names": null,
    "checksum": "                                                                ",
    "crate": "foo_vers_show_no_pb",
    "crate_size": 0,
    "created_at": "[datetime]",
    "dl_path": "https://crates.io/api/v1/crates/foo_vers_show_no_pb/1.0.0/download",
    "downloads": 0,
    "features": {},
    "has_lib": null,
    "id": "[id]",
    "lib_links": null,
    "license": null,
    "links": {
      "authors": "https://crates.io/api/v1/crates/foo_vers_show_no_pb/1.0.0/authors",
      "dependencies": "https://crates.io/api/v1/crates/foo_vers_show_no_pb/1.0.0/dependencies",
      "version_downloads": "https://crates.io/api/v1/crates/foo_vers_show_no_pb/1.0.0/downloads"
    },
    "num": "1.0.0",
    "published_by": null,
    "readme_path": "https://crates.io/api/v1/crates/foo_vers_show_no_pb/1.0.0/readme",
    "rust_version": null,
    "updated_at": "[datetime]",
    "yanked": false
  }
}
